src/main/scala/com/microsoft/hyperspace/index/CoveringIndex.scala

```diff
@@ -0,0 +1,420 @@
+      }
+
+      this.indexName = indexName
+      this
+    }
+
+    /**
+     * Updates column names for [[CoveringIndexConfig]].
+     *
+     * Note: API signature supports passing one or more argument.
+     *
+     * @param indexedColumn indexed column for the [[CoveringIndexConfig]].
+     * @param indexedColumns indexed columns for the [[CoveringIndexConfig]].
+     * @return an [[CoveringIndexConfig.Builder]] object with updated indexed columns.
+     */
+    def indexBy(indexedColumn: String, indexedColumns: String*): Builder = {
+      if (this.indexedColumns.nonEmpty) {
+        throw new UnsupportedOperationException("Indexed columns are already set.")
+      }
+
+      this.indexedColumns = indexedColumn +: indexedColumns
+      this
+    }
+
+    /**
+     * Updates included columns for [[CoveringIndexConfig]].
+     *
+     * Note: API signature supports passing one or more argument.
+     *
+     * @param includedColumn included column for [[CoveringIndexConfig]].
+     * @param includedColumns included columns for [[CoveringIndexConfig]].
+     * @return an [[CoveringIndexConfig.Builder]] object with updated included columns.
+     */
+    def include(includedColumn: String, includedColumns: String*): Builder = {
+      if (this.includedColumns.nonEmpty) {
+        throw new UnsupportedOperationException("Included columns are already set.")
+      }
+
+      this.includedColumns = includedColumn +: includedColumns
+      this
+    }
+
+    /**
+     * Creates IndexConfig from supplied index name, indexed columns and included columns
+     * to [[CoveringIndexConfig.Builder]].
+     *
+     * @return an [[CoveringIndexConfig]] object.
+     */
+    def create(): CoveringIndexConfig = {
+      CoveringIndexConfig(indexName, indexedColumns, includedColumns)
+    }
+  }
+
+  /**
+   *  Creates new [[CoveringIndexConfig.Builder]] for constructing an [[CoveringIndexConfig]].
+   *
+   * @return an [[CoveringIndexConfig.Builder]] object.
+   */
+  def builder(): Builder = new Builder
+}
+
+case class CoveringIndexMetadata(
+    config: CoveringIndexConfig,
+    indexedColumns: Seq[String],
+    includedColumns: Seq[String],
+    // see schemaJson for more information about the annotation
+    @JsonDeserialize(converter = classOf[StructTypeConverter.T]) schema: StructType,
+    numBuckets: Int,
+    properties: Map[String, String])
+    extends IndexMetadata {
+
+  override def indexName: String = config.indexName
+
+  override def withNewProperties(newProperties: Map[String, String]): IndexMetadata = {
+    copy(properties = newProperties)
+  }
+
+  override def withResolvedColumns(columnResolver: ColumnResolver): IndexMetadata = {
+    copy(
+      indexedColumns = columnResolver(config.indexedColumns),
+      includedColumns = columnResolver(config.includedColumns))
+  }
+
+  override def createIndex(): CoveringIndex = {
+    CoveringIndex(this)
+  }
+
+  override def createIndexData(
```

Comment:
Something seems wrong to this class... The case class is called `CoveringIndexMetadata` but it is mainly modifying the index data (aka the index data frame) and not the index metadata (aka the log entries).

Reply:
Please refer to the new API. I removed IndexMetadata and made it simpler.